crates/bevy_reflect/src/serde/de/deserializer.rs

```diff
@@ -14,6 +14,157 @@ use crate::{
+/// #     fn with_asset_type_id(&mut self, (): ()) -> &mut Self { unimplemented!() }
+/// #     fn untyped(&mut self) -> &mut Self { unimplemented!() }
+/// #     fn load_asset(&mut self, (): ()) -> Handle<LoadedUntypedAsset> { unimplemented!() }
+/// # }
+/// #
+/// # struct ReflectHandle;
+/// # impl TypeData for ReflectHandle {
+/// #     fn clone_type_data(&self) -> Box<dyn TypeData> { unimplemented!() }
+/// # }
+/// # impl ReflectHandle {
+/// #     fn asset_type_id(&self) { unimplemented!() }
+/// # }
+/// #
+/// #[derive(Debug, Clone, Reflect)]
+/// struct MyAsset {
+///     name: String,
+///     mesh: Handle<Mesh>,
+/// }
+///
+/// # type AssetError = Box<dyn core::error::Error>;
+/// # fn load<AssetPathVisitor: for<'de> Visitor<'de, Value = ()> + Copy>(
+/// #     asset_bytes: &[u8],
+/// #     type_registry: &TypeRegistry,
+/// #     load_context: &mut LoadContext,
+/// #     AssetPathVisitor: AssetPathVisitor,
+/// # ) -> Result<MyAsset, AssetError> {
+/// let mut ron_deserializer = ron::Deserializer::from_bytes(asset_bytes)?;
+/// let mut processor = ReflectDeserializerProcessor::new(
+///     |registration: &TypeRegistration| registration.data::<ReflectHandle>().is_some(),
+///     |registration, deserializer| {
+///         let reflect_handle = registration.data::<ReflectHandle>().unwrap();
+///         let asset_type_id = reflect_handle.asset_type_id();
+///
+///         let asset_path = deserializer.deserialize_str(AssetPathVisitor)?;
+///
+///         // takes in `load_context` from its scope
+///         let handle: Handle<LoadedUntypedAsset> = load_context
+///             .load()
+///             .with_asset_type_id(asset_type_id)
+///             .untyped()
+///             .load_asset(asset_path);
+/// #       let _: Result<_, ()> = {
+///         Ok(Box::new(handle))
+/// #       };
+/// #       unimplemented!()
+///     },
+/// );
+/// let reflect_deserializer =
+///     ReflectDeserializer::new_with_processor(type_registry, &mut processor);
+/// let asset = reflect_deserializer.deserialize(&mut ron_deserializer)?;
+/// # unimplemented!()
+/// # }
+/// ```
+///
+/// [`can_deserialize`]: Self::can_deserialize
+/// [`deserialize`]: Self::deserialize
+pub struct ReflectDeserializerProcessor<'p> {
+    /// When deserializing a value of which we know the type (via
+    /// [`TypeRegistration`]), do we want this processor to override the default
+    /// deserialization?
+    ///
+    /// If this returns [`true`], [`deserialize`] is used to create the
+    /// reflected value.
+    ///
+    /// [`deserialize`]: Self::deserialize
+    pub can_deserialize: Box<dyn FnMut(&TypeRegistration) -> bool + 'p>,
+    /// Deserializes a value for which [`can_deserialize`] returned [`true`].
+    ///
+    /// If you potentially return [`Ok`], you must consume the deserializer,
+    /// even if you don't use its output, otherwise the deserializer will be
+    /// in an invalid state.
```

Comment:
Is there a reason we can't give full ownership to this function?

Reply:
We could pass it `Box<dyn erased_serde::Deserializer>`, but:
- this means that we would need to unnecessarily box here:
```diff
            // First, check if our processor wants to deserialize this type
            // This takes priority over any other deserialization operations
            if let Some(processor) = self.processor.as_deref_mut() {
                if (processor.should_deserialize)(self.registration) {
-                  let mut deserializer = <dyn erased_serde::Deserializer>::erase(deserializer);
+                  let deserializer = Box::new(<dyn erased_serde::Deserializer>::erase(deserializer));
                    return (processor.deserialize)(self.registration, &mut deserializer)
                        .map_err(make_custom_error);
                }
            }
```
- this isn't actually useful, since `erased_serde::Deserializer` fns all take `&mut self`. There's no benefit in taking ownership

If you're asking because of the "you must consume the deserializer" doc, this doesn't actually refer to dropping (we can't drop anyway we only have a `&mut dyn`), but refers instead to the `deserializer.deserialize_ignored_any` that you have to do.

I ran into this while adding unit tests, but if you don't `deserialize` at least *something* here (even if you ignore its output), then you get situations like this:
```rs
(
  foo: 123,
  bar: 123,
)
```
fails to parse because "Expected comma". This detail should probably be expanded upon in the docs.